### PR TITLE
Qwen 2.5 VL Optimizations

### PR DESCRIPTION
**Pull request by arginugaTT**
### Ticket
This ticket contains all Qwen2.5 VL optimizations

### Problem description
To improve vision and text model performance of Qwen 2.5 VL

### What's changed
1. Batched Window Attention for vision block - https://github.com/tenstorrent/tt-metal/pull/37866
2. Batched Prefill optimization for text decoder, reference PR: https://github.com/tenstorrent/tt-metal/pull/35722
3. Vision Matmul Tuning
4. Minimal Matmul for Long Sequences (experimental)

### Solution Description
1. Batched Prefill

> Qwen2.5-VL's `prefill_forward_text` loops for user_id in range(batch) and calls `__prefill_forward_single_user_text` sequentially. When the inference server has 32 concurrent users, each user is prefilled one at a time.
> 
> - When all users have the same padded seq_len and batch <= 32, concatenate their tokens into a single tensor [1, 1, batch * seq_len, hidden_dim]
> - Process all users in one forward pass through the model
> - Loop only for KV cache fill (per-user page tables)
> - Reshape outputs back per user
> 
> Qwen2.5-VL uses custom M-RoPE. The RoPE cos/sin matrices are computed per-user by the processor (with vision-aware positions). In batched mode, we'd need to concatenate per-user RoPE matrices along the sequence dimension.

2. Vision Tuning

> `self.xqkv_prefill_progcfg = lambda seq_len: ttnn.MatmulMultiCoreReuseMultiCastProgramConfig(
>     compute_with_storage_grid_size=(8, 8),
>     in0_block_w=1,  # FIX: optimize this config for prefill
>     out_subblock_h=1,
>     out_subblock_w=1,`
> 
> Increase in0_block_w to reduce iterations. Tune other parameters & check if L1 memory fits etc.

3. Minimal Matmul

> For sequences >= 4096, standard ttnn.linear is suboptimal. Galaxy models use ttnn.experimental.minimal_matmul with tuned MinimalMatmulConfig for MLP.
> The MinimalMatmulConfig is tuned for Galaxy's core grid (7, 9). N300 has a different core grid (8, 8). Configs need retuning. This is exploratory. 

### Checklist

- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:arg/qwen_optimizations)

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:arg/qwen_optimizations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:arg/qwen_optimizations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:arg/qwen_optimizations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:arg/qwen_optimizations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:arg/qwen_optimizations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:arg/qwen_optimizations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:arg/qwen_optimizations)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:arg/qwen_optimizations)

### CODE REVIEW

**Comment by arginugaTT** [comment]
Performance Stats on tt-metal: (window attention + batched prefill)

image

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -80,7 +80,14 @@ def _set_sampling_trace_mode(self, enabled: bool):
...
 
-    def warmup_model_prefill(self, kv_cache, enable_trace, can_sample_on_device, non_greedy_decoding_on_device):
+    def warmup_model_prefill(
+        self,
+        kv_cache,
+        enable_trace,
+        can_sample_on_device=None,
+        non_greedy_decoding_on_device=None,
+        sampling_params=None,
+    ):
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: function signature should not be modified. sampling_params arg was removed recently and all remaining args are required.

File: models/demos/qwen25_vl/tt/generator.py [hunk 2]
```diff
@@ -217,13 +372,63 @@ def read_decode_output(self, tt_out, async_read=False):
...
-        logger.warning("Warmup model prefill not implemented for Qwen2_5_VL Generator")
-        logger.warning("Tracing in prefill mode is not supported for Qwen2_5_VL")
+    def warmup_model_prefill(
+        self,
+        kv_cache,
+        enable_trace,
+        can_sample_on_device=None,
+        non_greedy_decoding_on_device=None,
+        sampling_params=None,
+    ) -> None:
```
Comment (skhorasganiTT) [inline models/demos/qwen25_vl/tt/generator.py]: function signature should not be modified. sampling_params arg was removed recently and all remaining args are required.

File: models/demos/qwen25_vl/tt/generator.py [hunk 3]
```diff
@@ -217,13 +372,63 @@ def read_decode_output(self, tt_out, async_read=False):
...
+        """
+        Pre-compile programs for expected power-of-2 prefill sequence lengths.
+        Uses dummy embeddings and rotation matrices to avoid depending on real data.
+        """
+
```
Comment (skhorasganiTT) [inline models/demos/qwen25_vl/tt/generator.py:630]: sampling params need to be properly created and used by this function via:
```
sampling_params = self._create_sampling_params(
                            can_sample_on_device=can_sample_on_device,
                            non_greedy_decoding_on_device=non_greedy_decoding_on_device,
                            batch_size=batch_size,
                        )
```
see models/tt_transformers/tt/generator.py for reference.

File: models/demos/qwen25_vl/tt/generator_vllm.py [hunk 4]
```diff
@@ -19,6 +19,24 @@
 )
 from vllm.multimodal import MULTIMODAL_REGISTRY
 
+try:
+    import vllm.envs as envs
+
+    _VLLM_USE_V1 = envs.VLLM_USE_V1
```
Comment (skhorasganiTT) [inline models/demos/qwen25_vl/tt/generator_vllm.py]: V0 has been deleted from vLLM and is no longer supported. VLLM_USE_V1 should not be used anywhere and we shouldn't add any more v0 tech debt (this PR is adding back V0 code that was intentionally deleted recently).

File: models/demos/qwen25_vl/tt/generator.py [hunk 5]
```diff
@@ -2,12 +2,15 @@
```
Comment (skhorasganiTT) [inline models/demos/qwen25_vl/tt/generator.py:2]: Will qwen3 vl be updated similarly in this PR or a follow up PR? Also should they share generators/utilities if they're nearly identical?

**Review by mtairum — CHANGES_REQUESTED** [review]
Which Qwen2.5-VL are these changes targeting? 7B only?

If so, this is not yet part of the 3-tier system, correct? https://github.com/tenstorrent/tt-metal/blob/main/tests/pipeline_reorg/models_e2e_tests.yaml

Let me know if that's the case and I can review properly.

Otherwise if it's part of the 3-tier system or you want to make it part of, please add the test there accordingly and run those CI.

We don't have a document on how to migrate, but you'll need to add the test to:
- https://github.com/tenstorrent/tt-metal/blob/main/tests/pipeline_reorg/models_e2e_tests.yaml
- https://github.com/tenstorrent/tt-metal/blob/main/tests/pipeline_reorg/models_unit_tests.yaml
- Relevant Tier pipeline (either tier 2 or tier 3): 
   -  e.g. tier 2 e2e: https://github.com/tenstorrent/tt-metal/actions/workflows/models-t2-e2e-tests.yaml
   - e.g. tier 2 unit: https://github.com/tenstorrent/tt-metal/actions/workflows/models-t2-unit-tests.yaml 
- Doc: https://github.com/tenstorrent/tt-metal/blob/main/models/model_ci_tiers.md#tier-1-models
- 

Again, not necessary for this PR, if this test is not yet part of the 3-tier system and we do not wish to migrate at this point in time. Let me know.

**Comment by mtairum** [comment]
Any updates @arginugaTT ?

**Comment by arginugaTT** [comment]
> Any updates @arginugaTT ?

I will test other variants with same my changes and then run the CIs and merge this ticket @mtairum

**Comment by yieldthought** [comment]
Codex review:

[models/demos/qwen25_vl/tt/generator.py:397](https://github.com/tenstorrent/tt-metal/blob/f77a648c5096a6d3dc99e7469e094916fd024eab/models/demos/qwen25_vl/tt/generator.py#L397)

The new batched-prefill path drops the slot IDs that vLLM passes in `empty_slots`. `prefill_forward_text` accepts `empty_slots`, but when `use_batched_prefill` is true it slices compact request rows and `__prefill_forward_batched_text` rebuilds `batch_user_ids` as `range(batch_size)`. Those IDs are passed to `_get_prefill_user_page_table(..., use_batched_prefill=True)` and then to attention, where `user_id` is used as the concrete cache slot list for `paged_fill_cache(..., batch_idx=slot_idx)`. If vLLM refills slots such as `[5, 9]` while earlier requests are still active, this path writes the new users' K/V into slots 0 and 1, and with DP it can also choose the wrong shard because grouping is based on compact request indices instead of slot IDs. Decode for the scheduled slots then reads stale or another request's cache. Please preserve the per-request slot IDs through the batched path (or disable batched prefill unless `empty_slots == list(range(batch))`), scatter tokens/rot-mats/page-table rows to those slots like the shared generator does, pass the local slot IDs as `user_id`, and gather logits back to compact request order.
